### PR TITLE
feat(review): [R20 S4] triage promotion — never auto-approve risk-bearing dbt metadata

### PR DESCRIPTION
**Pull request by sahrizvi**
## Summary

Adds triage-tier promotion rules to `risk-tier.ts` so the reviewer never auto-approves risk-bearing dbt metadata changes. Two failure modes closed:

- Test-only YAML under `models/marts/` adding `data_tests:` / `constraints:` / `unique_combination_of_columns` on a contracted mart → previously classified `trivial` → auto-approved.
- SQL change under a FinOps-relevant path (cost / saving / billing / dbu / spend / …) → previously classified `lite` → approved without the deeper detector lanes running.

Three new `FileChangeClass` signals wire into `fullTierReasons()`:

| Signal | Promotes on its own? | Fires when |
|---|---|---|
| `dbtRiskYmlChanges` | Yes | schema.yml diff at YAML key position (not comments / description strings) adds/edits `data_tests:` / `constraints:` / `contract:` / `unique_combination_of_columns` list item |
| `martLayerChange` | **No** — enrichment only | File under `models/marts/` or `models/mart/`. Description-only edits under marts stay trivial (matches existing behavior). |
| `finopsPathToken` | Yes | Path/filename contains `cost\|saving\|billing\|credit\|dbu\|spend\|revenue\|price\|rate\|pricing\|invoice` at word/segment/extension boundary. FP guard on incidental substrings (`broadcaster` ≠ `caste`). |

## Test plan

- [x] 11 new regression tests in `packages/opencode/test/altimate/review.test.ts` — all green.
- [x] Existing 85 tests unchanged.
- [x] Full altimate review suite: 3781 pass / 640 skip / 0 fail (132 files).
- [x] Independently code-reviewed — two highs addressed (regex tightening + FileChangeClass consumer audit).

## Ship criteria

- [x] The two target failure modes must no longer auto-approve.
- [x] Composite Baseline B recall must not regress from 14/64 on the 13-scenario corpus.

## Depends on

Stacks on top of #1027 (`feat/review-r18-observability-recall`). This branch uses the `tierReasons[]` wiring landed there so promotion reasons surface in the signed envelope. Merge #1027 first, then rebase this onto main.

## Summary by CodeRabbit

* **New Features**
  * Added optional path-token rules (`riskTierPathTokens`) to promote matching changes to the full review risk tier.
  * Introduced preset-based token lists (including FinOps-related presets) with `preset:<name>` expansion.
  * Enhanced risk classification for `schema.yml` by detecting specific high-risk YAML keys and model-layer context.
* **Bug Fixes**
  * Reduced false positives from YAML block-scalar contents, comments, diff-header noise, and interior/partial word matches.
  * Unknown presets now degrade gracefully, recording a configuration error in risk tier reasons.
* **Tests**
  * Expanded risk-tier coverage for key matching, token boundaries, preset expansion, and config validation.

### CODE REVIEW

**Comment by sahrizvi** [comment]
## Validation report against offline dbt-PR corpus

Ran the branch CLI against a set of merged dbt PRs used for offline validation (both `--pure --no-ai=true` and `--pure` with the LLM lane). Baseline is the parent PR's CLI on the same corpus.

### Recall delta
| Variant | Baseline findings | Branch findings | Delta |
|---|---:|---:|---:|
| `--pure --no-ai=true` (deterministic only) | 118 | 118 | **+0** |
| `--pure` (LLM lane on) | 141 | 143 | +2 (LLM run-to-run noise) |

### Tier changes
One PR in the corpus was previously classified `lite`; this branch correctly re-classifies it to `full` via the FinOps-path signal. The other four were already `full` for other reasons (blast radius / pre-existing `touchesContract`), so this branch doesn't change their tier.

### Interpretation

The change is a **routing** improvement, not a detector. On this corpus:
- ✅ The under-tiered PR now correctly runs at `full` tier.
- ✅ `tierReasons[]` surfaces explicit grounded reasons (mart-API surface, FinOps keyword, data_tests/constraints).
- ⚠️ Zero deterministic finding-count improvement — the extra full-tier lanes (contract_violation, pii_exposure, materialization, warehouse_cost, idempotency) didn't have anything to fire on for that PR's specific diff shape.
- ⚠️ +2 LLM delta is run-to-run noise, not attributable to this branch.

**Value is foundational**: this branch unblocks subsequent detector PRs from being neutered by miscalibrated triage. Real recall wins show up in the detector-work PRs stacked above it.

**Comment by sahrizvi** [comment]
# Consensus Code Review — PR #1028

**Title:** feat(review): [S4] triage promotion — never auto-approve risk-bearing dbt metadata
**Branch:** `feat/review-r20-s4-triage-promotion` → `feat/review-r18-observability-recall` (stacks on #1027)  ·  **Repo:** AltimateAI/altimate-code
**Size:** 2 files, +202 (`risk-tier.ts` +53, `review.test.ts` +149)  ·  **Reviewed:** 2026-07-22

**Panel:** multi-model review.

---

## Verdict: REQUEST CHANGES — 2 MAJOR (targeted fixes), 5 MINOR, 3 NIT

The promotion mechanism is verified correct by all 8 reviewers (`fullReasons` is checked before the trivial/lite branches, so the new signals genuinely force `full` and can never slip to auto-approve). The change is well-scoped; the blockers are a false-negative gap and a vacuous test, both targeted fixes.

**Counts:** Critical 0 · Major 2 · Minor 5 · Nit 3

---

## MAJOR

### 1. Legacy `tests:` key is not promoted → re-opens the auto-approve path it aims to close
**Category:** False-negative / tiering safety · **Location:** `risk-tier.ts` `DBT_RISK_KEY_RE`
The regex covers `data_tests | constraints | contract` but NOT the pre-dbt-1.8 `tests:` key. A schema.yml adding `tests: [- unique / - not_null]` under `models/marts/`, with no FinOps token or blast radius, falls through `onlyDocs` → `tier: "trivial"` → auto-approve — the exact failure mode this PR targets, for projects on dbt <1.8 or mid-migration. Partial mitigation: `unique_combination_of_columns` under legacy `tests:` is still caught by the key-agnostic `DBT_UNIQUE_COMBO_RE`, but plain `unique`/`not_null`/`relationships`/`accepted_values` under `tests:` are not.
**Fix:** `(?:tests|data_tests|constraints|contract)` + a regression test proving legacy `tests:` promotes. If pre-1.8 is out of scope, document that explicitly.
_8-way consensus (top finding, unanimous in convergence). Original severities ranged NIT→CRITICAL; settled at MAJOR._

### 2 → downgraded to MINOR in convergence (see MINOR section)

### 3. The `unique_combination_of_columns` regression test is vacuous (test conflation)
**Category:** Test correctness · **Location:** `review.test.ts` "full: schema.yml adds unique_combination_of_columns test"
The test path `models/marts/<column>.yml` also matches `FINOPS_TOKEN_RE` (`billing`, `prices`) and sits under `marts/`, so `finopsPathToken` promotes it to `full` independently of the diff — the assertion `tier === "full"` passes even if `DBT_UNIQUE_COMBO_RE` were deleted. The one test meant to lock in unique-combo detection can never fail on that regex.
**Fix:** move to a non-FinOps, non-mart path (e.g. `models/intermediate/int_grain.yml`), keep the diff, and assert both `tier === "full"` and the reason string, so `DBT_UNIQUE_COMBO_RE` is the sole possible promoter.
_a reviewer (unique catch); lead confirmed `billing`+`prices` both match at the shell. Convergence: <reviewer> + <reviewer> + a reviewer hold MAJOR; A reviewer dissents (MINOR — test-only, no production impact). Kept MAJOR: a core detector's only regression test providing false confidence is meaningful for a review-tool codebase._

---

## MINOR

### 2. FinOps token over-promotes on non-financial analytics paths *(downgraded from MAJOR in convergence)*
**Category:** Cost / precision · **Location:** `risk-tier.ts` `FINOPS_TOKEN_RE` + `finopsPathToken` hard floor
`rate | rates | pricing | credit | saving | dbus` match at `_`/`.`/`-` boundaries on non-FinOps paths — verified at the shell: `mrt_error_rate.sql`, `stg_fill_rate.sql`, `mrt_conversion_rate.sql`, `mrt_retention_rate.sql` (true), and `stg_dbus_connector.sql` (D-Bus IPC, not Databricks Units — true). Because `finopsPathToken` is a hard floor before the size/blast gates, any change to these files runs full tier (LLM cost).
**Why MINOR (convergence):** multiple reviewers, and <reviewer> independently argued this is a deliberate recall-over-precision fail-safe — it never under-promotes or creates an auto-approve regression; it only adds review cost. That is a precision/cost tradeoff, not a tiering-safety defect.
**Fix (optional):** narrow `rate`/`rates`/`pricing` (require financial context or couple with marts/reporting), drop `dbus` (keep `dbu`), and/or gate path-only promotion by non-doc changed lines.
_Note: an earlier example (`stg_accredited_loans.sql`, "`credit` inside `accredited`") was **retracted** — A reviewer flagged it and the shell confirms it does NOT match (no boundary char around `credit`)._

### 4. Promotion reason can omit the actual triggering signal
**Category:** Observability · The reason always reads "schema.yml diff touches data_tests/constraints/contract" even when the match came solely from `unique_combination_of_columns`, so the signed `tierReasons` misnames the trigger. Name the exact signal (or return structured signal names). _GPT + a reviewer_

### 5. "upgrades the WEIGHT" comment is misleading
`martLayerChange` only appends a location string to the reason; there is no actual weight/ordering change (the tier is already `full`). Fix the comment or implement real weighting. _A reviewer._

### 6. Block-scalar / multi-line description false positive
A YAML block scalar (`|`/`>`) whose continuation line begins with `data_tests:`/`constraints:` matches `DBT_RISK_KEY_RE`; the `(?!#)` guard only excludes `#` comments, not block-scalar bodies. Low frequency. _a reviewer_

### 7. FinOps boundary misses digit suffixes (false-negative)
`mrt_cost2024.sql`, `stg_billing2.sql`, `dbu1_usage.sql` don't match — the char after the keyword is a digit, not in `[\/_.-]`. Add `\d` to the boundary class. _A reviewer._

---

## NIT
8. `DBT_UNIQUE_COMBO_RE` requires a trailing `:` and list-item marker → misses the bare/indented-key short form. _Claude, <reviewer>, a reviewer_
9. Removed-line (`- data_tests:` / `- constraints:`) risk keys promote — correct (removing a guardrail is risk-worthy) but untested. _MiMo, a reviewer_
10. Comment calls `unique_combination_of_columns` a "TEST NAME"; it's a test macro/parameter. _A reviewer._

---

## Rejected findings (for transparency)
- **"No dedicated `contract:` promotion test"** (multiple reviewers MINOR) — **false**: `review.test.ts:405` tests `+ contract:\n+ enforced: true`. A reviewer flagged this; lead confirmed. Removed.
- **"FINOPS_TOKEN_RE lacks `/i`"** (<reviewer> NIT) — false; the regex carries `/i` (risk-tier.ts:84).
- **"Mart `.yml` may classify as `other`, missing `dbtRiskYmlChanges`"** (<reviewer> MAJOR) — false; `classifyDbtFile` routes any `.yml` under `models/` to `schema_yml` (diff-filter.ts:75).
- **"`martLayerChange` does nothing / `fullTierReasons` empty unless pushed"** (<reviewer> CRITICAL) — that is the intended enrichment-only design, confirmed by every other reviewer and the tests.

## Positive observations (consensus)
- Promotion ordering verified correct by all 8 reviewers: `fullReasons` checked before trivial/lite → `dbtRiskYmlChanges`/`finopsPathToken` reliably force `full`; never slips to auto-approve.
- `martLayerChange` correctly enrichment-only (never promotes alone) — description-only mart edits stay trivial; tested.
- Diff-level detection (`changedLines`) prevents surrounding context from promoting.
- Good FP guards + negative tests: comment/description exclusion, incidental substrings (`broadcaster`, `precast`), and `MARTS_DIR_RE` requiring `marts/` as a path segment (rejects `models/marts_backup/`, `mymodels/marts/`).

## Missing tests
- Legacy `tests:` promotion (guards MAJOR no. 1). · A non-FinOps/non-mart path for the unique-combo test (guards MAJOR no. 3). · Removed-line risk key promotes (NIT no. 9). · Optional: FinOps over-promotion acceptance test for `*_rate` analytics paths (MINOR no. 2).

---

## Convergence

| Reviewer | Verdict on synthesis | Key point |
|---|---|---|
| a reviewer | CHANGES NEEDED → incorporated | no. 2 → MINOR (cost-only fail-safe); `accredited` example wrong; `contract:` test does exist (removed MINOR no. 8); no. 3 is MINOR/low-MAJOR. |
| a reviewer | CHANGES NEEDED → incorporated | no. 2 → MINOR; no. 3 correctly MAJOR; no. 1 correctly top. |
| a reviewer | CHANGES NEEDED → incorporated | no. 2 → MINOR (`dbus` is a real bug though); no. 3 correctly MAJOR. |
| Reviewer | (author) | Agreed with all three downgrades/corrections. |

**Applied in convergence:** no. 2 MAJOR→MINOR (3 reviewers, unanimous); retracted the `accredited` example (shell-verified non-match); removed the "no `contract:` test" MINOR (test exists at review.test.ts:405). no. 3 kept MAJOR (<reviewer> + <reviewer> + a reviewer vs <reviewer>'s lone MINOR). no. 1 unchanged as top finding. Converged in 1 round.

## Finding attribution

| # | Finding | Origin | Type |
|---|---------|--------|------|
| 1 | Legacy `tests:` key not promoted | multiple reviewers, a reviewer, multiple reviewers | Consensus (8-way) |
| 3 | Vacuous `unique_combination_of_columns` test | a reviewer | Unique (shell-verified) |
| 2 | FinOps over-promotion on `*_rate`/`dbus` paths | multiple reviewers, a reviewer, multiple reviewers | Consensus; severity set MINOR in convergence |
| 4 | Reason omits actual triggering signal | <reviewer>, a reviewer | Unique |
| 5 | "upgrades the WEIGHT" comment misleading | multiple reviewers | Unique |
| 6 | Block-scalar description FP | a reviewer | Unique |
| 7 | FinOps boundary misses digit suffixes | Reviewer | Unique |
| 8–10 | NITs (short-form regex, removed-line untested, comment) | multiple reviewers, a reviewer, <reviewer> | Nit |
| ✓ | Promotion ordering correct; martLayerChange enrichment-only | all 8 | Consensus (positive) |

*Reviewed by a multi-model panel.*

File: packages/opencode/src/altimate/review/risk-tier.ts [hunk 1]
```diff
@@ -60,11 +122,80 @@ function changedLines(diff: string | undefined): string {
...
+ *  added line inside its body contains `data_tests:` etc. Tracking scalar
+ *  state over the whole diff — and only masking output on changed lines —
+ *  closes that gap while preserving the +/- filter's role of ignoring
+ *  surrounding-code noise. */
+function changedLinesForScan(diff: string | undefined): string {
+  if (!diff) return ""
+  const lines = diff.split("\n")
+  const out: string[] = []
+  let scalarIndent = -1
+  const blockScalarStart = /^[ \t]*[^\s#:][^:]*:[ \t]*[|>][+-]?[ \t]*$/
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/altimate/review/risk-tier.ts]: Addressed in `e52c863ae`. `blockScalarStart` regex widened to `[|>](?:[+-]?[1-9]?|[1-9]?[+-]?)[ \t]*(?:#.*)?$` so `|2`, `|+2`, `|2-` (explicit indentation indicator, either order relative to chomp) and `| # comment` (trailing comment) all open a scalar. Regression test loops over `|2`, `|+2`, `| # legacy comment` shapes and asserts each produces trivial on a diff whose body contains risk keywords.

File: packages/opencode/src/altimate/review/risk-tier.ts [hunk 2]
```diff
@@ -60,11 +122,88 @@ function changedLines(diff: string | undefined): string {
...
+  let scalarIndent = -1
+  // YAML block-scalar header. The optional trailing tail covers:
+  //  - explicit indentation indicator (1-9) either before or after the
+  //    chomping indicator: `|2`, `|+2`, `|2+`, `>2-`
+  //  - trailing comment: `description: | # legacy`
+  // kilo-code-bot suggestion — the earlier form matched only `|`/`>`
+  // with an optional `+`/`-` chomp, so `|2` and `| # comment` skipped the
+  // opener → body lines beginning with a risk keyword false-positive
+  // promoted (safe-direction over-tiering, but still wrong).
+  const blockScalarStart = /^[ \t]*[^\s#:][^:]*:[ \t]*[|>](?:[+-]?[1-9]?|[1-9]?[+-]?)[ \t]*(?:#.*)?$/
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/altimate/review/risk-tier.ts:232]: Addressed in `cd8b9709e`. The marker/stripped logic now strips the leading diff marker on ALL lines (context = leading space, changed = `+`/`-`) before indent measurement, so context and changed lines live in the same coordinate system. New regression test: a `|1` block scalar opened in context with body containing `data_tests:` / `constraints:` prose stays trivial.

File: packages/opencode/test/altimate/review.test.ts [hunk 3]
```diff
@@ -501,6 +501,361 @@ describe("risk-tier", () => {
...
+      expect(r.tier).toBe("full")
+      expect(r.reasons.join(" ")).toContain("FinOps keyword")
+    }
+  })
+
+  test("R20 S4: FinOps token does NOT over-fire on incidental substrings", () => {
+    // False-positive guard: the token regex requires path/word boundaries so
+    // words like `broadcaster.py` (has `caste`) or `precast_table.sql` (has
+    // `cast`) don't fire the cost/dbu/etc. rules.
+    const r1 = classifyPR([file("models/staging/stg_broadcaster.sql", "+select 1\n")], {
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/test/altimate/review.test.ts]: Addressed — the strip commit (`21f99146e`) moves the FinOps token list out of `risk-tier.ts` core into user-configurable `.altimate/review.yml` (`riskTierPathTokens`), keeps the shipped list as an opt-in `preset:finops`, and replaces this vacuous test with property-based loops over `RISK_TOKEN_PRESETS.finops`:

- **Positive (fires):** `stg_${tok}_daily.sql` for every preset token — the boundary case.
- **Negative (does NOT fire):** `stg_x${tok}y.sql` — letters on both sides so the alternation matches only because of the interior substring; only the boundary anchors block promotion. If someone weakens the boundary class, every one of these fails simultaneously.
- **Neutrality guarantee:** `no promotion when no resolver is configured (core is neutral)` — a project that doesn't opt into any category gets zero path-token promotion.

Adding a preset token now extends coverage automatically. Thanks for the catch.

File: packages/opencode/src/altimate/review/risk-tier.ts [hunk 4]
```diff
@@ -45,11 +68,117 @@ export interface ClassifyOptions {
...
+    "invoices",
+  ],
+}
+
+/**
+ * Compile a `riskTierPathTokens` config record into a `pathTokenCategoryOf`
+ * resolver — used by orchestration to inject the callback into `classifyPR`.
+ * Handles `preset:<name>` expansion. Tokens match at path/word/digit
+ * boundaries: `mrt_cost.sql` fires, `broadcaster.sql` doesn't. Returns
+ * undefined (no resolver) when no categories are configured, so the
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/altimate/review/risk-tier.ts:113]: Design call — the opt-in default is intentional, not an oversight. The strip's premise is that the reviewer *core* must be neutral about which business verticals (billing / PCI / patient-safety / anything else) a project cares about; hardcoding one team's FinOps vocabulary into every consumer's core is what the change was undoing. The trade-off you flag (S4's default protection reverts unless a project opts in) is accepted — projects that want the FinOps promotion enable it in one line: `riskTierPathTokens: {finops: [preset:finops]}` in `.altimate/review.yml`. Shipping the preset default-on would re-leak the vertical into core (same argument that motivated the strip). This is an explicit scope change from the original S4 ship criteria; noting for the merge log.

File: packages/opencode/src/altimate/review/config.ts [hunk 5]
```diff
@@ -46,6 +46,29 @@ export const ReviewConfig = z.object({
...
+   *   riskTierPathTokens:
+   *     pci: [card, pan, cvv]
+   *     patient: [phi, mrn, diagnosis]
+   *
+   * Matching is case-insensitive at path/word/digit boundaries — a token
+   * `cost` fires on `mrt_cost_daily.sql` and `mrt_cost2024.sql` but not
+   * inside `broadcaster.sql` or `precast_table.sql`. The `preset:<name>`
+   * marker expands to the current shipped list at reviewer startup.
+   */
+  riskTierPathTokens: z.record(z.string(), z.array(z.string())).default({}),
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/altimate/review/config.ts]: Fixed in `29c242323` — `riskTierPathTokens: z.record(z.string(), z.array(z.string().min(1)))`. Regression test added under `describe("config")` in `review.test.ts` asserts throw on empty token + non-empty tokens still parse. Thanks for the catch.

File: packages/opencode/src/altimate/review/risk-tier.ts [hunk 6]
```diff
@@ -60,11 +189,93 @@ function changedLines(diff: string | undefined): string {
...
+    }
+    // Only added/removed lines are candidates for regex scanning; context
+    // lines only feed scalar-state tracking. When the changed line lives
+    // inside a scalar, emit a blank so the regex has nothing to match.
+    if (marker === " ") continue
+    out.push(insideScalar ? "" : raw)
+  }
+  return out.join("\n")
+}
+
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/altimate/review/risk-tier.ts:274]: Valid — traced the code.

At line 228 `marker` is `"+"`, `"-"`, `" "`, or `""` (fallback for anything else). At line 246 the guard only skips `marker === " "`, so a line with an empty marker — `diff --git ...`, `index abc..def`, `Author: ...`, `Date: ...` — falls through and gets pushed to `out`, then feeds `dbtRiskYmlKeyMatches`.

Real-world exposure right now is near-zero: `file.diff` from GitHub's PR files API arrives as raw hunks starting at `@@`, so this never hits in production. Would any DBT risk-key regex actually match a git-header line? Not by construction — the patterns require the exact word (`data_tests`/`tests`/`constraints`/`contract`) at YAML key position followed by `:`, and no header text collides. But the code was asymmetric: `+++`/`---`/`@@` unified-diff headers were explicitly skipped at line 220, while other free-form lines weren't. A future caller passing raw `git diff -p` output would silently see those in the scanner.

Fixed in `9b2c024e1` — swapped the guard to `if (marker !== "+" && marker !== "-") continue`. Any line whose leading character isn't `+` or `-` is either a context line (fine, already updated scalar state above) or a free-form header (skipped, no emit). Added a regression test `raw git diff -p free-form headers do NOT reach the scanner` — builds a full raw-diff string (four header lines + hunk header + one changed line) and asserts the same tier verdict as the bare-`tests`-word test. 231/231 review-* tests pass.

File: packages/opencode/src/altimate/review/risk-tier.ts [hunk 7]
```diff
@@ -45,11 +68,117 @@ export interface ClassifyOptions {
...
+  cfg: Record<string, string[]>,
+): ((path: string) => string | undefined) | undefined {
+  const compiled: Array<{ category: string; re: RegExp }> = []
+  for (const [category, entries] of Object.entries(cfg)) {
+    const tokens: string[] = []
+    for (const entry of entries) {
+      if (entry.startsWith("preset:")) {
+        const name = entry.slice("preset:".length)
+        const preset = RISK_TOKEN_PRESETS[name]
+        if (preset) tokens.push(...preset)
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/altimate/review/risk-tier.ts]: Fixed in `919bcc17f` — `compilePathTokenResolver` now throws on an unknown `preset:<name>` before assembling the resolver:

```
Error: riskTierPathTokens.finops: unknown preset 'finop'.
Known presets: finops. Configure a bare token list (e.g.
['card', 'pan']) instead if you want a custom category.
```

Rationale: the shipped preset list is small and stable, so an unknown name is almost certainly a typo — not an intentional forward reference. Fail at review-start means the misconfiguration is caught immediately, not discovered when a real high-risk PR slips through unchecked.

Regression test `unknown preset name throws with the known-list in the message` asserts throw on `preset:finop`, throw when a typo is mixed with valid tokens in the same category (so a partially-broken resolver is never returned), and that `preset:finops` still resolves promotion correctly on `mrt_cost_daily.sql`.

File: packages/opencode/src/altimate/review/risk-tier.ts [hunk 8]
```diff
@@ -45,11 +68,117 @@ export interface ClassifyOptions {
...
+      } else {
+        tokens.push(entry)
+      }
+    }
+    if (tokens.length === 0) continue
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/altimate/review/risk-tier.ts:147]: Fixed in `919bcc17f` — `compilePathTokenResolver` now throws on an unknown `preset:<name>` before assembling the resolver:

```
Error: riskTierPathTokens.finops: unknown preset 'finop'.
Known presets: finops. Configure a bare token list (e.g.
['card', 'pan']) instead if you want a custom category.
```

Rationale: the shipped preset list is small and stable, so an unknown name is almost certainly a typo — not an intentional forward reference. Fail at review-start means the misconfiguration is caught immediately, not discovered when a real high-risk PR slips through unchecked.

Regression test `unknown preset name throws with the known-list in the message` asserts throw on `preset:finop`, throw when a typo is mixed with valid tokens in the same category (so a partially-broken resolver is never returned), and that `preset:finops` still resolves promotion correctly on `mrt_cost_daily.sql`.

**Review by saravmajestic — DISMISSED** [review]

File: packages/opencode/src/altimate/review/risk-tier.ts [hunk 9]
```diff
@@ -45,11 +68,133 @@ export interface ClassifyOptions {
...
+        const preset = RISK_TOKEN_PRESETS[name]
+        // Unknown preset name is almost certainly a typo (e.g.
+        // `preset:finop` instead of `preset:finops`); the shipped list is
+        // small and stable. Silently dropping the entry means the user's
+        // opt-in for a risk-promotion category quietly does nothing —
+        // the exact safety invariant this PR exists to guarantee. Fail
+        // loud instead so the misconfiguration is caught at review
+        // start, not discovered when a real high-risk PR slips through
+        // (cubic-review P2 + altimate-harness-bot on PR #1028).
+        if (!preset) {
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/altimate/review/risk-tier.ts:134]: Fixed in `9435bc6f9` — `orchestrate.ts` now wraps the `compilePathTokenResolver` call in a try/catch, so a single config typo no longer crashes the run. On throw:

1. Prominent warning to **stderr** with the resolver's error message + `Review continuing without path-token promotion` fallback banner.
2. Fall back to `pathTokenCategoryOf = undefined` so `classifyPR` still runs — no path-token promotion fires, but the rest of the review proceeds normally.
3. Prepend the config error to `tierResult.reasons` so it surfaces in `envelope.tierReasons` when `explainTier` is set — a reader who never sees stderr (envelope JSON, PR-comment renderer) still knows why their opt-in didn't fire.

Trade-off is deliberate: the user's opt-in is dead until they fix the typo (safe fallback — no silent auto-approve on billing/PCI paths), but the review itself doesn't die.

Regression test `runReview does NOT crash on an unknown-preset config typo — degrades gracefully`: builds a config with `finops: [preset:finop]`, captures stderr, asserts (a) envelope is produced (no crash), (b) stderr contains the config-error message + fallback banner, (c) `env.tierReasons` carries the same error.

File: packages/opencode/src/altimate/review/orchestrate.ts [hunk 10]
```diff
@@ -1092,14 +1092,41 @@ export async function runReview(input: OrchestrateInput): Promise<VerdictEnvelop
...
       const c = ctxByPath.get(p)
       return c ? c.impact.directCount + c.impact.transitiveCount : 0
     },
     touchesPiiOf: (f) => (ctxByPath.get(f.path)?.pii.length ?? 0) > 0,
     isComplexOf: (f) => ctxByPath.get(f.path)?.complex ?? false,
+    pathTokenCategoryOf,
   })
+  // Surface config errors in the tier-reasons stream so a reader of the
+  // envelope (or PR comment) sees WHY their opt-in didn't fire.
+  if (pathTokenConfigError) tierResult.reasons.unshift(pathTokenConfigError)
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/altimate/review/orchestrate.ts:1129]: Fixed in `3c2c03c99`. Expanded the envelope gate:

```ts
tierReasons: input.explainTier || tierForced || pathTokenConfigError
             ? tierReasons
             : undefined
```

Now the config-error path always surfaces in the envelope (and downstream PR comment), regardless of `--explain-tier`. Regression test `config error surfaces in envelope even WITHOUT --explain-tier` calls `runReview` with a `preset:finop` typo AND `explainTier` unset, asserts `env.tierReasons` still contains both the config-invalid message and the exact typoed preset name. Thanks for the catch — real silent-failure gap in my earlier fix.

**Review by dev-punia-altimate — APPROVED** [review]
reviewing as requested
